### PR TITLE
[2/N] Add -Wdeprecated and related fixes

### PR DESCRIPTION
**Pull request by cyyever**
This PR follows #108626.
cc @ezyang @Skylion007

### CODE REVIEW

File: aten/src/ATen/core/TensorBase.h [hunk 1]
```diff
@@ -205,6 +205,7 @@ class TORCH_API TensorBase {
     impl_.reset();
   }
 
+#if defined (_MSC_VER)
```
Comment (ezyang) [inline aten/src/ATen/core/TensorBase.h:208]: oh??!

Comment (cyyever) [reply aten/src/ATen/core/TensorBase.h:208]: MSVC will break on 
```
  TensorBase& operator=(const TensorBase& x) & = default;
  TensorBase& operator=(TensorBase&& x) & noexcept = default;
```

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**